diff --git a/src/future.rs b/src/future.rs
index 30c0323..49325b4 100644
--- a/src/future.rs
+++ b/src/future.rs
@@ -65,7 +65,7 @@ mod private_try_future {
 
 /// A convenience for futures that return `Result` values that includes
 /// a variety of adapters tailored to such futures.
-pub trait TryFuture: Future + private_try_future::Sealed {
+pub trait TryFuture: Future<Output = Result<<Self as TryFuture>::Ok, <Self as TryFuture>::Error>> + private_try_future::Sealed {
     /// The type of successful values yielded by this future
     type Ok;
 
diff --git a/src/stream.rs b/src/stream.rs
index dd07d5a..4495c61 100644
--- a/src/stream.rs
+++ b/src/stream.rs
@@ -174,7 +174,7 @@ mod private_try_stream {
 
 /// A convenience for streams that return `Result` values that includes
 /// a variety of adapters tailored to such futures.
-pub trait TryStream: Stream + private_try_stream::Sealed {
+pub trait TryStream: Stream<Item = Result<<Self as TryStream>::Ok, <Self as TryStream>::Error>> + private_try_stream::Sealed {
     /// The type of successful values yielded by this future
     type Ok;
 
